File: core/MaximumNumberOfAchievableTransferRequests/MaximumNumberOfAchievableTransferRequests.java
package com.dbc;

import java.util.Arrays;
import java.util.List;
import java.util.Stack;

public class MaximumNumberOfAchievableTransferRequests {
    private int[] remain = new int[20];
    private int maxScores = 0;
    private int[][] requests;

    private boolean isBalance() {
        for (int item : this.remain) {
            if (item > 0) return false;
        }
        return true;
    }

    private void dfs(Stack<Integer> reqList, int start) {
        for (int i = start; i < this.requests.length; i++) {
            reqList.push(i);
            remain[this.requests[i][0]]--;
            remain[this.requests[i][1]]++;
            if (isBalance()) {
                this.maxScores = Math.max(this.maxScores, reqList.size());
            }
            dfs(reqList, i + 1);
            reqList.pop();
            remain[this.requests[i][0]]++;
            remain[this.requests[i][1]]--;
        }
    }

    public int maximumRequests(int n, int[][] requests) {
        this.requests = requests;
        dfs(new Stack<>(), 0);
        return this.maxScores;
    }
}

File: core/MaximumNumberOfAchievableTransferRequests/MaximumNumberOfAchievableTransferRequests.py
from typing import List


def maximumRequests(n: int, requests: List[List[int]]) -> int:
    floor = {i: 0 for i in range(n)}
    length = len(requests)
    max_req = 0

    def dfs(req_list, start):
        nonlocal max_req, floor
        for index in range(start, length):
            req_list.append(index)
            floor[requests[index][0]] -= 1
            floor[requests[index][1]] += 1
            if sum([val for val in floor.values() if val > 0]) == 0:
                max_req = max(max_req, len(req_list))
            dfs(req_list, index + 1)
            req_list.pop()
            floor[requests[index][0]] += 1
            floor[requests[index][1]] -= 1

    dfs([], 0)
    return max_req


if __name__ == '__main__':
    print(maximumRequests(n = 5, requests = [[0,1],[1,0],[0,1],[1,2],[2,0],[3,4]]))
